src/cosl/coordinated_workers/nginx.py

```diff
@@ -29,6 +31,425 @@
+        resolv = Path(RESOLV_CONF_PATH).read_text()
+        for line in resolv.splitlines():
+            if line.startswith("nameserver"):
+                # assume there's only one
+                return line.split()[1].strip()
+        raise RuntimeError("cannot find nameserver in /etc/resolv.conf")
+
+    def _upstreams(self, addresses_by_role: Dict[str, Set[str]]) -> List[Any]:
+        nginx_upstreams: List[Any] = []
+
+        for upstream_config in self._upstream_configs:
+            addresses = addresses_by_role.get(upstream_config.worker_role)
+            # don't add an upstream block if there are no addresses
+            if addresses:
+                nginx_upstreams.append(
+                    {
+                        "directive": "upstream",
+                        "args": [upstream_config.name],
+                        "block": [
+                            {"directive": "server", "args": [f"{addr}:{upstream_config.port}"]}
+                            for addr in addresses
+                        ],
+                    }
+                )
+
+        return nginx_upstreams
+
+    def _build_servers_config(
+        self, backends: List[str], tls: bool = False
+    ) -> List[Dict[str, Any]]:
+        servers: List[Dict[str, Any]] = []
+        for port, locations in self._server_ports_to_locations.items():
+            server_config = self._build_server_config(
+                port,
+                locations,
+                backends,
+                tls,
+            )
+            if server_config:
+                servers.append(server_config)
+        return servers
+
+    def _build_server_config(
+        self,
+        port: int,
+        locations: List[NginxLocationConfig],
+        backends: List[str],
+        tls: bool = False,
+    ) -> Dict[str, Any]:
+        auth_enabled = False
+        is_grpc = any(loc.is_grpc for loc in locations)
+        nginx_locations = self._locations(locations, is_grpc, backends, tls)
+        server_config = {}
+        if len(nginx_locations) > 0:
+            server_config = {
+                "directive": "server",
+                "args": [],
+                "block": [
+                    *self._listen(port, ssl=tls, http2=is_grpc),
+                    *self._basic_auth(auth_enabled),
+                    {
+                        "directive": "proxy_set_header",
+                        "args": ["X-Scope-OrgID", "$ensured_x_scope_orgid"],
+                    },
+                    {"directive": "server_name", "args": [self._server_name]},
+                    *(
+                        [
+                            {"directive": "ssl_certificate", "args": [CERT_PATH]},
+                            {"directive": "ssl_certificate_key", "args": [KEY_PATH]},
+                            {
+                                "directive": "ssl_protocols",
+                                "args": ["TLSv1", "TLSv1.1", "TLSv1.2", "TLSv1.3"],
```

Comment:
From my point of view, these TLS protocol versions are important enough to extract them as modules constants. This way it is much more clear to know about the supported versions at glance.